tools/engine_tool/lib/src/gn_utils.dart

```diff
@@ -0,0 +1,117 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:io';
+
+import 'package:path/path.dart' as p;
+import 'package:process_runner/process_runner.dart';
+
+import 'environment.dart';
+import 'proc_utils.dart';
+
+/// Information about a test build target.
+final class TestTarget {
+  /// Construct a test target.
+  TestTarget(this.label, this.executable);
+
+  /// The build target label.
+  final String label;
+
+  /// The executable file produced after the build target is built.
+  final File executable;
+
+  @override
+  String toString() {
+    return 'target=$label executable=${executable.path}';
+  }
+}
+
+/// Returns test targets for a given build directory.
+Future<Map<String, TestTarget>> findTestTargets(
+    Environment environment, Directory buildDir) async {
+  final Map<String, TestTarget> r = <String, TestTarget>{};
+  final List<String> getLabelsCommandLine = <String>[
+    'gn',
+    'ls',
+    buildDir.path,
+    '--type=executable',
+    '--testonly=true',
+    '--as=label',
+  ];
+  final ProcessRunnerResult labelsResult = await environment.processRunner
+      .runProcess(getLabelsCommandLine,
+          workingDirectory: environment.engine.srcDir, failOk: true);
+  fatalIfFailed(environment, getLabelsCommandLine, labelsResult);
+  final String rawLabels = labelsResult.stdout;
+  final List<String> getOutputsCommandLine = <String>[
+    'gn',
+    'ls',
+    buildDir.path,
+    '--type=executable',
+    '--testonly=true',
+    '--as=output'
+  ];
+  final ProcessRunnerResult outputsResults = await environment.processRunner
+      .runProcess(getOutputsCommandLine,
+          workingDirectory: environment.engine.srcDir, failOk: true);
+  fatalIfFailed(environment, getOutputsCommandLine, outputsResults);
+  final String rawOutputs = outputsResults.stdout;
+  final List<String> labels = rawLabels.split('\n');
+  final List<String> outputs = rawOutputs.split('\n');
+  if (labels.length != outputs.length) {
+    environment.logger.fatal(
+        'gn ls output is inconsistent A and B should be the same length:\nA=$labels\nB=$outputs');
+  }
+  // Drop the empty line at the end of the output.
+  if (labels.isNotEmpty) {
+    assert(labels.last.isEmpty);
```

Comment:
FYI: Neither the `et` wrapper script nor the unit tests run with `--enable-asserts`. Do you want to add the flag to the unit tests in this PR?

Reply:
I would also be supportive of just making this an unconditional throw, since it's going to throw anyway.

Reply:
If we never run with --enable-asserts can we add a lint that will flag assert calls for us, or, start running with asserts enabled?

Also, made this a throw.